std/algorithm/iteration.d

```diff
@@ -4289,3 +4304,602 @@ unittest
+
+    assert(iota(0).cartesianPower.length == 0);
+    assert("AB"d.cartesianPower(3).equal!equal(["AAA"d, "AAB"d, "ABA"d, "ABB"d, "BAA"d, "BAB"d, "BBA"d, "BBB"d]));
+    auto expected = ["AA"d, "AB"d, "AC"d, "AD"d,
+                     "BA"d, "BB"d, "BC"d, "BD"d,
+                     "CA"d, "CB"d, "CC"d, "CD"d,
+                     "DA"d, "DB"d, "DC"d, "DD"d];
+    assert("ABCD"d.cartesianPower(2).equal!equal(expected));
+    // verify with array too
+    assert("ABCD"d.cartesianPower(2).map!array.array == expected);
+
+    assert(iota(2).cartesianPower.front.equal([0]));
+
+    // is copyable?
+    auto a = iota(2).cartesianPower;
+    assert(a.front.equal([0]));
+    assert(a.save.dropOne.front.equal([1]));
+    assert(a.front.equal([0]));
+}
+
+/**
+Lazily computes all k-combinations of $(D r).
+Imagine this as the $(LREF cartesianPower) filtered for only strictly ordered items.
+
+For example $(D"AB".combinations(2).array) returns $(D["AB"]).
+
+Params:
+    r = $(REF_ALTTEXT Random access range, isRandomAccessRange, std, range, primitives)
+    k = number of combinations
+
+Returns:
+    Forward range which yields the k-combinations items
+*/
+auto combinations(Range)(Range r, size_t k = 1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(k >= 1, "Invalid repeat");
+}
+body
+{
+    import std.array: array;
+    import std.range: Indexed;
+    import std.range: iota;
+    static struct Combination
+    {
+    private:
+        Range _r;
+        size_t[] _state;
+
+        size_t _repeat, _nrElements;
+        size_t _max_states, _pos;
+
+    public:
+        this(Range r, size_t repeat)
+        {
+            _r = r;
+            // cache number of elements
+            _nrElements = r.length;
+            _repeat = repeat;
+
+            // set initial state and calculate max possibilities
+            if (_nrElements > 0)
+            {
+                _max_states = binomial(_nrElements, _repeat);
+            }
+
+            if (_nrElements > 0 && _max_states > 0)
+            {
+                _state.length = repeat;
+                // skip first duplicate
+                if (_nrElements > 1 && _repeat > 1)
+                {
+                    _state = iota(repeat).array;
+                }
+            }
+        }
+
+        @property typeof(this) save()
+        {
+            // we need to save more information about the current state
+            auto c = typeof(this)(_r, _repeat);
+            c._state = _state.dup;
+            c._pos = _pos;
+            return c;
+        }
+
+        @property Indexed!(Range, size_t[]) front()
+        {
+            import std.range : indexed;
+            return _r.indexed(_state);
+        }
+
+        void popFront()
+        {
+            // check whether we have reached the end
+            if (empty) return;
+            _pos++;
+            // we might have bumped into the end state now
+            if (empty) return;
```

Comment:
If `popFront` is called for the last state, we run below zero at the `while` loop below.

Alternatives:

- `adding `i > 0` to the loop
- `if (  _pos == _max_states - 1) { _pos++; return; }`  

Imho the end case happens seldom, so we want to avoid it if possible and for the other too algorithms we can do that.